File: Algorithms/31. Next Permutation/swift/solution.swift
class Solution {
    func reverse(_ nums: inout [Int], _ left: Int, _ right: Int) {
        var left = left
        var right = right

        while left < right {
            nums.swapAt(left, right)
            left += 1
            right -= 1
        }
    }

    func nextPermutation(_ nums: inout [Int]) {
        guard nums.count > 1 else {
            return
        }

        var idx1 = nums.count - 2
        while idx1 >= 0 && nums[idx1] >= nums[idx1 + 1] {
            idx1 -= 1
        }

        if idx1 >= 0 {
            var idx2 = nums.count - 1
            while idx2 >= 0 && nums[idx2] <= nums[idx1] {
                idx2 -= 1
            }
            nums.swapAt(idx1, idx2)
        }
        reverse(&nums, idx1 + 1, nums.count - 1)
    }
}

File: Algorithms/31. Next Permutation/kotlin/solution.kt
class Solution {
    fun nextPermutation(nums: IntArray) {
        var i = nums.size - 2

        while (i >= 0 && nums[i + 1] <= nums[i]) {
            i--
        }

        if (i >= 0) {
            var j = nums.size - 1
            while (j >= 0 && nums[j] <= nums[i]) {
                j--
            }
            swap(nums, i, j)
        }
        reverse(nums, i + 1)
    }

    fun swap(nums: IntArray, i: Int, j: Int) {
        val temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp
    }

    fun reverse(nums: IntArray, i: Int) {
        var left = i
        var right = nums.size - 1
        while (left < right) {
            swap(nums, left, right)
            left++
            right--
        }
    }
}
